Android/Samples/libraries/kml/src/main/java/com/microsoft/maps/kml/KMLParser.java

```diff
@@ -79,46 +86,130 @@ private void parsePlacemark() throws IOException, XmlPullParserException, KMLPar
       }
       String type = mParser.getName();
       if (type.equals("Placemark")) {
-        // used for MapIcon Title
-        String name = parseName();
+        parseNameAndShape();
+      }
+    }
+  }
+
+  private void parseNameAndShape() throws IOException, XmlPullParserException, KMLParseException {
+    String title = null;
+    MapElement element = null;
+    while (mParser.next() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      switch (type) {
+        case "name":
+          mParser.require(XmlPullParser.START_TAG, mNameSpace, "name");
+          title = parseText();
+          mParser.require(XmlPullParser.END_TAG, mNameSpace, "name");
+          break;
+        case "Point":
+          element = parsePoint();
+          break;
+        default:
+          skip();
+      }
+    }
+    if (element != null) {
+      if (title != null && element instanceof MapIcon) {
+        ((MapIcon) element).setTitle(title);
       }
+      mLayer.getElements().add(element);
     }
   }
 
-  @Nullable
-  private String parseName() throws IOException, XmlPullParserException, KMLParseException {
+  @NonNull
+  private MapElement parsePoint() throws IOException, XmlPullParserException, KMLParseException {
+    mParser.require(XmlPullParser.START_TAG, mNameSpace, "Point");
     while (mParser.next() != XmlPullParser.END_TAG) {
       if (mParser.getEventType() != XmlPullParser.START_TAG) {
         continue;
       }
       String type = mParser.getName();
-      if (type.equals("name")) {
-        mParser.require(XmlPullParser.START_TAG, mNameSpace, "name");
-        String title = parseText();
-        mParser.require(XmlPullParser.END_TAG, mNameSpace, "name");
-        return title;
+      if (type.equals("coordinates")) {
+        ArrayList<Geoposition> coordinates = parseCoordinates();
+        if (coordinates.size() > 1) {
+          throw new KMLParseException(
+              "coordinates for a Point can only contain one position. Instead saw: "
+                  + coordinates.size()
+                  + " at position: "
+                  + mParser.getPositionDescription());
+        }
+        MapIcon icon = mFactory.createMapIcon();
+        icon.setLocation(new Geopoint(coordinates.get(0)));
```

Comment:
How do you know that this element exitsts?

Reply:
parseCoordinates returns a nonnull ArrayList of Geopositions. If no Geoposition could be created, an error would be thrown. I'll add a comment mentioning this requirement.

Reply:
Test case added for parseCoordinates catching an empty string as an invalid Geoposition.